adapters/handlers/grpc/v1/batch/queues.go

```diff
@@ -0,0 +1,452 @@
+//                           _       _
+// __      _____  __ ___   ___  __ _| |_ ___
+// \ \ /\ / / _ \/ _` \ \ / / |/ _` | __/ _ \
+//  \ V  V /  __/ (_| |\ V /| | (_| | ||  __/
+//   \_/\_/ \___|\__,_| \_/ |_|\__,_|\__\___|
+//
+//  Copyright © 2016 - 2025 Weaviate B.V. All rights reserved.
+//
+//  CONTACT: hello@weaviate.io
+//
+
+package batch
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"sync"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	enterrors "github.com/weaviate/weaviate/entities/errors"
+	pb "github.com/weaviate/weaviate/grpc/generated/protocol/v1"
+)
+
+type QueuesHandler struct {
+	shuttingDownCtx  context.Context
+	logger           logrus.FieldLogger
+	writeQueues      *WriteQueues
+	readQueues       *ReadQueues
+	sendWg           *sync.WaitGroup
+	streamWg         *sync.WaitGroup
+	shutdownFinished chan struct{}
+}
+
+func NewQueuesHandler(shuttingDownCtx context.Context, sendWg, streamWg *sync.WaitGroup, shutdownFinished chan struct{}, writeQueues *WriteQueues, readQueues *ReadQueues, logger logrus.FieldLogger) *QueuesHandler {
+	// Poll until the batch logic starts shutting down
+	// Then wait for all BatchSend requests to finish and close all the write queues
+	// Scheduler will then drain the write queues expecting the channels to be closed
+	ticker := time.NewTicker(100 * time.Millisecond)
+	enterrors.GoWrapper(func() {
+		for {
+			select {
+			case <-shuttingDownCtx.Done():
+				logger.Info("shutting down batch queues handler, waiting for in-flight requests to finish")
+				sendWg.Wait()
+				logger.Info("all in-flight requests finished, closing write queues")
+				writeQueues.Close()
+				logger.Info("write queues closed, exiting handlers shutdown listener")
+				return
+			case <-ticker.C:
+			}
+		}
+	}, logger)
+	return &QueuesHandler{
+		shuttingDownCtx:  shuttingDownCtx,
+		logger:           logger,
+		writeQueues:      writeQueues,
+		readQueues:       readQueues,
+		sendWg:           sendWg,
+		streamWg:         streamWg,
+		shutdownFinished: shutdownFinished,
+	}
+}
+
+func (h *QueuesHandler) wait(ctx context.Context) error {
+	ticker := time.NewTicker(100 * time.Millisecond)
+	for {
+		select {
+		case <-ctx.Done():
```

Comment:
to prevent resource leaks you should also stop a ticker here in this case: `ticker.Stop()`